src/library/basesqltablemodel.cpp

```diff
@@ -419,19 +375,102 @@ void BaseSqlTableModel::setSort(int column, Qt::SortOrder order) {
         qDebug() << this << "setSort()" << column << order;
     }
 
-    bool sortColumnChanged = m_iSortColumn != column;
-    bool sortOrderChanged = m_eSortOrder != order;
+    int trackSourceColumnCount = m_trackSource ? m_trackSource->columnCount() : 0;
 
-    if (!sortColumnChanged && !sortOrderChanged) {
-        // Do nothing if the sort is not different.
+    if (column < 0 ||
+            column >= trackSourceColumnCount + m_sortColumns.size() - 1) {
+        // -1 because id column is in both tables
+        qWarning() << "BaseSqlTableModel::setSort invalid column:" << column;
         return;
     }
 
-    // TODO(rryan) optimization: if the sort column has not changed but the
-    // order has, just reverse our ordering of the rows.
+    if (m_sortColumns.size() > 0 &&
+            m_sortColumns.at(0).column == column) {
+        // Only the order has changed
+        SortColumn sc;
+        sc.column = column;
+        sc.order = order;
+        m_sortColumns.replace(0, sc);
+    } else {
+        // remove column if already in history
+        for (int i = 0; i < m_sortColumns.size(); ++i) {
+            if (m_sortColumns.at(i).column == column) {
+                m_sortColumns.removeAt(i);
```

Comment:
if the first column is removed (i=0), then everything shifts to the left and this loop will skip the second item, which is now at i=0.